Frameworks/Foundation/NSOperationQueue.mm

```diff
@@ -1,443 +1,381 @@
+                [_operations removeObjectAtIndex:index];
+                [self didChangeValueForKey:@"operationCount"];
+                [self didChangeValueForKey:@"operations"];
             }
-            [priv->suspendedCondition unlock];
 
-            if (RunOperationFromLists(&priv->myQueues[i], &priv->queues[i], &priv->curOperation)) {
-                didWork = TRUE;
-            }
+            // Clean up state from addOperation:
+            [op removeObserver:self forKeyPath:@"isFinished" context:nil];
+            [op removeObserver:self forKeyPath:@"isReady" context:nil];
         }
-    } while (didWork);
-
-    [innerPool release];
-
-    return self;
+    } else if ([keyPath isEqualToString:@"isReady"]) {
+        NSOperation* op = static_cast<NSOperation*>(object);
+        [self _startOperation:op];
+    }
 }
 
 /**
  @Status Interoperable
 */
-- (BOOL)hasMoreWork {
-    if (priv->curOperation != nil) {
-        return TRUE;
++ (NSOperationQueue*)currentQueue {
+    NSThread* currentThread = [NSThread currentThread];
+    if ([currentThread isMainThread]) {
+        return [[self class] mainQueue];
     }
 
-    for (int i = 0; i < NSOperationQueuePriority_Count; i++) {
-        if (priv->queues[i] != NULL) {
-            return TRUE;
-        }
-    }
-
-    return FALSE;
+    return [[currentThread threadDictionary] objectForKey:_NSOperationQueueCurrentQueueKey];
 }
 
 /**
  @Status Interoperable
 */
-- (id)init {
-    if (self = [super init]) {
-        priv = new NSOperationQueuePriv();
-
-        priv->_maxConcurrentOperationCount = 1;
-        priv->workAvailable = [[NSCondition alloc] init];
-        priv->suspendedCondition = [[NSCondition alloc] init];
-        priv->allWorkDone = [[NSCondition alloc] init];
-        priv->isSuspended = 0;
-
-        _completionQueue = dispatch_queue_create("NSOperation finish queue", nullptr);
-    }
-
-    return self;
-}
-
-- (id)_initMainThread {
-    priv = new NSOperationQueuePriv();
-
-    priv->_maxConcurrentOperationCount = 1;
-    priv->workAvailable = [[NSCondition alloc] init];
-    priv->suspendedCondition = [[NSCondition alloc] init];
-    priv->allWorkDone = [[NSCondition alloc] init];
-    priv->isSuspended = 0;
-
-    return self;
++ (NSOperationQueue*)mainQueue {
+    static StrongId<NSOperationQueue> mainQueue = [[NSOperationQueue_MainQueue new] autorelease];
+    return mainQueue;
 }
 
 /**
  @Status Interoperable
 */
-- (void)addOperation:(id)op {
-    /*
-    //  Add any dependencies
-    id dependencies = [op dependencies];
-    int count = [dependencies count];
-    for ( int i = 0; i < count; i ++ ) {
-    id curDep = [dependencies objectAtIndex:i];
-    [self addOperation:curDep];
+- (void)addOperation:(NSOperation*)op {
+    if (op.isExecuting || op.isFinished) {
+        @throw [NSException exceptionWithName:NSInvalidArgumentException reason:@"operation is already or finished executing" userInfo:nil];
```

Comment:
We may want to have an exception formatter that prepends `-[ClassName selector:name:]` to our exceptions; i know we have a couple scattered around.